### ISSUE TITLE
nonode@nohost couch file epoch error

### ISSUE THREAD

**Issue by nickva** [apache/couchdb#5130]
Noticed one of the production nodes's `_dbs` database was throwing this error:

```
with exit value:#012{epoch_order,[
  {couch_db,validate_epochs,1,[{file,"src/couch_db.erl"},{line,1758}]},
  {couch_db,get_epochs,1,[{file,"src/couch_db.erl"},{line,584}]},
  {mem3_rep,compare_epochs,2,[{file,"src/mem3_rep.erl"},{line,557}]},
  ... {mem3_rep,go,1,[{file,"src/mem3_rep.erl"},{line,122}]}]}
```

Examining the db handle close noticed it has `nonode@nohost` epoch entries.

```
> (element(2,Db#db.engine))#st.header#db_header.epochs, couch_db:close(Db)
[{'actualname@actual.host.net',23436889},
 {nonode@nohost,23353728},
 {'actualname@actual.host.net',842473},
 {nonode@nohost,830216},
 {'actualname@actual.host.net',18423167},
 {nonode@nohost,18401097},
 {'actualname@actual.host.net',2227879},
 {nonode@nohost,2204657},
 {'actualname@actual.host.net',10924197},
 {nonode@nohost,10924195},
 {'actualname@actual.host.net',4548304},
 {nonode@nohost,4534552},
 {'actualname@actual.host.net',0}]
```

`nonode@nohost` is what we expect `node()` to return if the node doesn't have a `-name/-sname` defined, is standalone.

All the nodes we have in production have a `-node actualname@actual.host.net` in the vm.args file, so the expectation is that `node()` will always return that node name, which clearly doesn't always happen.

The fact that his occurred with the dbs metadata db, may point that it has something to do with a startup or shutdown state: there is a period of time when the network dist sub-system either hasn't start or is shut down and we're opening the couch file and updating it, and so `node()` would return `nonode@nohost`.

Another idea is perhaps net_kernel crashes and restarts, but in that case we might expect a full node restart.

If this is indeed an expected state, and our previous assumption was wrong, we'd have to handle.  Perhaps when we update/check epoch we have to also check init state and arguments and if the user did specify a node name we either ignore, crash or patch up the state.

Maybe a combination of `node()` and these results can be useful:
```
(foo@bar.net)5> init:get_status().
{started,started}
(foo@bar.net)6> proplists:get_value(name, init:get_arguments()).
["foo@bar.net"]
(foo@bar.net)7> proplists:get_value(sname, init:get_arguments()).
undefined
```

We also can't blindly always ignore a `nonode@nohost` epoch change in case the user moved a file from a stand-alone couch file to a clustered setup.

**Comment by nickva** [comment]
There are actually two issues here:
   * We end up with `nonode@nohost`, `node()` response when we have a `-name ...` argument. The assumption in our code base is that can't happen.
   * We record recorded an epoch that re-wound the update sequence backward  18423167 to 830216 . That is what triggered the error in the logs. We detect this condition when we read the sequence.

**Comment by rnewson** [comment]
```Returns the name of the local node. If the node is not alive, nonode@nohost is returned instead.```

We can ping the folks over there for details but the docs for `node()` don't inspire confidence.

**Comment by nickva** [comment]
We were able to model the `nonode@nohost` result from `nodes()` state by noticing that it happen when the `net_kernel` dist controller process is killed.  The VM will exit, but in the process of shutting down for a period of time `node()` returns `nonode@nohost`. The sequence rewind still remains a mystery, though.

So far we've put in couple of fixes related to this:
  * https://github.com/apache/couchdb/pull/5132 introduces a permanent node name that saves the initial `node()` in a persistent term and does value validations right after in `config`, then asserts during epoch updates that the sequence rewinds.
  * https://github.com/apache/couchdb/pull/5138 uses the persistent node name in a few more cases when there is a likelihood of the `nonode@nohost` affecting persistent data (checkpoints, shard maps, logs)
  
Closing the issue for now. We'll monitor to see if we notice any assertion triggers on sequence rewinds and re-open it then to continue working on it.
